kong/plugins/correlation-id/handler.lua

```diff
@@ -0,0 +1,56 @@
+-- Copyright (C) Mashape, Inc.
+
+local BasePlugin = require "kong.plugins.base_plugin"
+local uuid = require "lua_uuid"
+local req_set_header = ngx.req.set_header
+local req_get_headers = ngx.req.get_headers
+
+local CorrelationIdHandler = BasePlugin:extend()
+
+local worker_uuid
+local worker_counter
+
+local generators = setmetatable({
+	["uuid"] = function()
+    return uuid()
+	end,
+	["uuid#counter"] = function()
+    worker_counter = worker_counter + 1
+    return worker_uuid.."#"..worker_counter
+	end,
+}, { __index = function(self, generator)
+    ngx.log(ngx.ERR, "Invalid generator: "..generator)
+end
+})
+
+function CorrelationIdHandler:new()
+  CorrelationIdHandler.super.new(self, "correlation-id")
+end
+
+function CorrelationIdHandler:init_worker()
+  CorrelationIdHandler.super.init_worker(self)
+  worker_uuid = uuid()
+  worker_counter = 0
+end
+
+function CorrelationIdHandler:access(conf)
+  CorrelationIdHandler.super.access(self)
+
+  -- Set header for upstream
+  local header_value = req_get_headers()[conf.header_name]
+  if not header_value then
+    -- Generate the header value
+    header_value = generators[conf.generator]()
+    req_set_header(conf.header_name, header_value)
+  end
+
+  -- For later use, to echo it back downstream
+  ngx.ctx.correlationid_header_value = header_value
```

Comment:
Echoing back downstream?
Sure it will - it will help the caller "correlate" if there was a problem with the request.
E.g. mockbin also echos `x-request-id` back:

```
curl http://mockbin.com/request --data 'foo=bar'

{
  "startedDateTime": "2016-03-22T19:42:26.954Z",
  "clientIPAddress": "77.101.29.24",
  "method": "POST",
  "url": "http://mockbin.com/request",
  "httpVersion": "HTTP/1.1",
  "cookies": {},
  "headers": {
    "host": "mockbin.com",
    "connection": "close",
    "accept-encoding": "gzip",
    "x-forwarded-for": "77.101.29.24, 162.158.152.161",
    "cf-ray": "287c2079a0a73578-LHR",
    "x-forwarded-proto": "http",
    "cf-visitor": "{\"scheme\":\"http\"}",
    "user-agent": "curl/7.47.1",
    "accept": "*/*",
    "x-request-id": "b04bad2e-c240-43ad-a0f9-60c98d86599d",
    "content-type": "application/x-www-form-urlencoded",
    "cf-connecting-ip": "77.101.29.24",
    "x-forwarded-port": "80",
    "via": "1.1 vegur",
    "connect-time": "0",
    "x-request-start": "1458675746943",
    "total-route-time": "0",
    "content-length": "7"
  },
  "queryString": {},
  "postData": {
    "mimeType": "application/x-www-form-urlencoded",
    "text": "foo=bar",
    "params": {
      "foo": "bar"
    }
  },
  "headersSize": 535,
  "bodySize": 7
}
```

Please confirm if you disagree, then I'll make it configurable.